### ISSUE TITLE
Improper Access Control by Directory Listing Misconfiguration

### ISSUE THREAD

**Issue by emaragkos** [gunet/openeclass#39]
Improper Access Control by Directory Listing Misconfiguration that affects all versions

When the webapp is poorly configured (directory listing is enabled), an unauthenticated user will be able to view and download students' uploaded assessments.

PoC:
Course link
https://127.0.0.1/courses/CS101
Add "work" directory at the end
https://127.0.0.1/courses/CS101/work/

**Comment by iamaldi** [comment]
Hi, 

I was wondering if there's a responsible disclosure policy that you and other researchers can follow instead of disclosing security issues in the public?

GitHub has the Security panel added to every repository but seems like this one is not configured yet.

Thanks

**Comment by Arisstath** [comment]
It doesn't seem to be a problem of the software itself. There are many ways that a misconfigured web server could lead to unauthorized access to various section of eclass.

**Comment by emaragkos** [comment]
That is true but the web application itself should ensure that in case of misconfiguration, no sensitive data is exposed through directory listing. It is a bad practice and the fact that this misconfiguration is pretty common kinda confirms it. Many institutions unfortunately are still affected by this issue and will continue to be unless a software fix is implemented. 

https://owasp.org/www-community/Broken_Access_Control

**Comment by XhmikosR** [comment]
@adia: has this been taken care of in v3.10.x? The changelog is a little cryptic...

**Comment by adia** [comment]
Hello, and thanks for reminding us of this bug report. You are right, this is still a problem, especially for smaller installations on shared hosting where directory listings might be enabled and admins might not notice the problem. We do advise disabling them in [our installation instructions](https://docs.openeclass.org/en/install).

We were preparing a new release (3.11) for tomorrow, and we'll make sure this is addressed by adding an empty "index.html" in all new subdirectories and creating it on upgrade in case it doesn't exist.

As for a disclosure policy, we'll publish something more official in GitHub's Security panel, but if you find any vulnerabilities, please report them by email to eclass@gunet.gr (which is read only by the core team).

Thanks again, and sorry for not taking care of your report earlier.

**Comment by XhmikosR** [comment]
@adia: please make sure you mention it properly in the release notes and also have someone update the CVE details. The report here was a valid one IMHO and this has been in public for so long...

If this is fixed, you might have to release new patch versions for 3.9.x etc; I'm unsure what's your support policy.

On a side note, at least the Bootstrap version you are using has known security issues, specifically quite a few XSS issues. You should update to the latest 3.4.1. jQuery and probably other plugins have further security issues too.

**Comment by adia** [comment]
Will do about giving credit and upgrading our JS dependencies - that "your report" was addressed to everyone participating here, but of course, thanks and credit goes to @emaragkos. Although - we knew that if directory listings are enabled, no files are secure, and that's why disabling directory indexes is one of the first installation steps we recommend.

The proper fix is to support putting the data directory outside the web root, so that there's no need for configuration on the web server to limit direct access, but given our scant development resources we haven't prioritized that. We'll try to work on this for the next release.

Our policy is that all installations should upgrade to the latest release, but if anyone asks, and offers a justification for not being able to upgrade (e.g. being in the middle of exams period so that any changes are unwelcome) we work with them to provide fixes for known problems, as long as the changes are localized - for wider changes we always recommend upgrading to the latest release since we can't properly test big back-ported changes properly.

**Comment by XhmikosR** [comment]
The thing is that right now even search engines have crawled plenty of files in some cases (the one I have access on). So, even if you place an empty `index.html` file or the organization fixes their bogus installation, the files are still accessible.

**Comment by emaragkos** [comment]
This issue might seem of little importance at first, but given the fact that during covid19 quarantine many institutions used eClass as a solution for online remote exams, it poses a huge impact on confidentiality of private and sensitive data.
It does not only expose students' personal data (name, surname, email, student ID, even photos with gov or academic ID used for proof of identity) but also exposes private semester assessments, unreleased exam subjects and submitted answers resulting in compromising the integrity of the whole examination process.
I think this should be fixed asap with a temporary patch such as the index.html mentioned until there is time for a proper solution.
I'm sorry for posting it here in the first place, I wasn't aware of the recommended alternative way to report similar issues.
